### PR TITLE
Fix Go major version updates

### PR DESCRIPTION
**Pull request by EyalDelarea**
- [x] All [tests](https://github.com/jfrog/frogbot#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
---

**Bug Description:**
Module paths in GO must have a major version suffix like /v2 that matches the major version.
For example, if a module has the `path example.com/mod at v1.0.0`, it must have the path` example.com/mod/v2` at version v2.0.0.
This PR makes sure this functionality is implemented.

Also fixed the bug where FrogBot suggested downgraded fixed versions

### CODE REVIEW

File: commands/createfixpullrequests.go [hunk 1]
```diff
@@ -307,42 +310,54 @@ func (cfp *CreateFixPullRequestsCmd) updatePackageToFixedVersion(packageType cor
...
-// commandName - Name of the package manager
-// commandArgs - Package manager upgrade command
+// technology - Name of the package manager
 // impactedPackage - Vulnerable package to upgrade
 // fixVersion - The version that fixes the vulnerable package
-// operator - The operator between the impactedPackage to the fixVersion
-func fixPackageVersionGeneric(commandName string, commandArgs []string, impactedPackage, fixVersion, operator string) error {
+func fixPackageVersionGeneric(technology coreutils.Technology, impactedPackage, fixVersion string) error {
+
+	if technology == coreutils.Go {
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:335]: Wouldn't it be better tp place this code inside a new `fixPackageVersionGo` to remain consistent with tthe way the code is currently structured? We have `fixPackageVersionMaven` and a few other similar functions. Why should we break the code symetry?

File: commands/createfixpullrequests.go [hunk 2]
```diff
@@ -307,42 +310,54 @@ func (cfp *CreateFixPullRequestsCmd) updatePackageToFixedVersion(packageType cor
...
 // In other cases, we had to handle the upgrade process
-// commandName - Name of the package manager
-// commandArgs - Package manager upgrade command
+// technology - Name of the package manager
 // impactedPackage - Vulnerable package to upgrade
 // fixVersion - The version that fixes the vulnerable package
-// operator - The operator between the impactedPackage to the fixVersion
-func fixPackageVersionGeneric(commandName string, commandArgs []string, impactedPackage, fixVersion, operator string) error {
+func fixPackageVersionGeneric(technology coreutils.Technology, impactedPackage, fixVersion string) error {
+
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:334]: Redundant new-line.

File: commands/createfixpullrequests.go [hunk 3]
```diff
@@ -307,42 +310,54 @@ func (cfp *CreateFixPullRequestsCmd) updatePackageToFixedVersion(packageType cor
...
-		commandArgs := []string{"install"}
-		err = fixPackageVersionGeneric(packageType.GetExecCommandName(), commandArgs, impactedPackage, fixVersion, "==")
 	case coreutils.Poetry:
 		err = fixPackageVersionPoetry(impactedPackage, fixVersion)
 	default:
-		return fmt.Errorf("package type: %s is currently not supported", string(packageType))
+		err = fixPackageVersionGeneric(packageType, impactedPackage, fixVersion)
+		if err != nil {
+			return err
+		}
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:323]: This if statement seems redundant, beicase the fucntion will return the value of `err` right after anyway.

File: commands/createfixpullrequests.go [hunk 4]
```diff
@@ -307,42 +310,54 @@ func (cfp *CreateFixPullRequestsCmd) updatePackageToFixedVersion(packageType cor
...
 	fixedPackage := impactedPackage + operator + fixVersion
+
 	commandArgs = append(commandArgs, fixedPackage)
-	return runPackageMangerCommand(commandName, commandArgs)
+	return runPackageMangerCommand(technology.GetExecCommandName(), commandArgs)
+}
+
+// Module paths in GO must have a major version suffix like /v2 that matches the major version.
+// For example, if a module has the path example.com/mod at v1.0.0, it must have the path example.com/mod/v2 at version v2.0.0.
+func formatMajorVersionSuffix(impactedPackage, fixVersion string) (string, error) {
```
Comment (eyalbe4) [inline commands/createfixpullrequests.go:352]: formatMajorVersionSuffix --> formatGoVersion

Let's also add a unit test for this function. You can use "table testing". You can see existing examples in the code for this.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
I'd like to review this PR again following the impelemnation of my inline comments.
Please reassign this PR to me when you're ready.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
Note that I changed the label for this PR from "improvement" to "bug".
Please go ahhead and add a clear description for this PR, explaining the bug it fixes. This si important, release notes are generated automatically from our PRs.

File: commands/createfixpullrequests.go (hunk 1 above)
Comment (EyalDelarea) [reply commands/createfixpullrequests.go:335]: I've looked and there is only one other use to this function. (in the poetry function)
wouldn't it be more generic if you didn't have to pass in the operator and the command name? 
instead you pass in the tech which has all of the constants.

File: commands/createfixpullrequests_test.go [hunk 5]
```diff
@@ -208,3 +208,40 @@ func verifyTechnologyNaming(t *testing.T, scanResponse []services.ScanResponse,
...
+			description:     "Validate no string has been added",
+		},
+	}
+
+	for _, tt := range formatGoCases {
+		t.Run(tt.description, func(t *testing.T) {
+			result, _ := formatGoVersion(tt.impactedPackage, tt.fixVersion)
+			if result != tt.expected {
+				t.Errorf("expected %s, but got %s", tt.expected, result)
+			}
```
Comment (omerzi) [inline commands/createfixpullrequests_test.go:244]:
```suggestion
			result, err := formatGoVersion(tt.impactedPackage, tt.fixVersion)
			assert.NoError(t, err)
			assert.Equal(t, tt.expected, result)
```

File: go.mod [hunk 6]
```diff
@@ -86,16 +85,18 @@ require (
...
+	golang.org/x/term v0.5.0 // indirect
+	golang.org/x/text v0.7.0 // indirect
+	golang.org/x/time v0.1.0 // indirect
 	google.golang.org/appengine v1.6.7 // indirect
 	google.golang.org/protobuf v1.28.1 // indirect
 	gopkg.in/ini.v1 v1.67.0 // indirect
 	gopkg.in/warnings.v0 v0.1.2 // indirect
 )
+
+replace github.com/jfrog/jfrog-cli-core/v2 => github.com/EyalDelarea/jfrog-cli-core/v2 v2.0.0-20230223153256-cc80c8ac3fe4
```
Comment (omerzi) [inline go.mod:102]: Don't forget to remove this line and update `jfrog-cli-core` to the relevant version after your changes there were merged.

**Review by omerzi — CHANGES_REQUESTED** [review]

**Review by eyalbe4 — APPROVED** [review]

**Comment by sverdlov93** [comment]
Hi @EyalDelarea,
An important thing you should consider is that the following change should be applied when dealing with 'github.com' dependencies only.
For example in gopkg:  `gopkg.in/yaml.v3 v3.0.1` they use `.` instead of `/`, and maybe there are more conventions.

File: commands/testdata/projects/go/hello.go [hunk 7]
```diff
@@ -2,12 +2,14 @@ package main
...
 	"fmt"
+	"github.com/go-git/go-git"
 	"github.com/google/uuid"
 	"github.com/sassoftware/go-rpmutils"
 )
 
 func main() {
 	fmt.Println("test")
 	uuid.New()
+	_ = git.Refs{}
```
Comment (sverdlov93) [inline commands/testdata/projects/go/hello.go:13]: Don't forget to remove it before merge

File: commands/createfixpullrequests.go [hunk 8]
```diff
@@ -467,3 +485,11 @@ func (fvi *FixVersionInfo) UpdateFixVersion(newFixVersion string) {
 		fvi.fixVersion = newFixVersion
 	}
 }
+
+func fixPackageVersionGo(impactedPackage, fixVersion string) error {
+	impactedPackage, err := formatGoVersion(impactedPackage, fixVersion)
```
Comment (sverdlov93) [inline commands/createfixpullrequests.go:490]:
```suggestion
	impactedPackage, err := handleGoPackageSemanticVersionSuffix(impactedPackage, fixVersion)
```

File: commands/createfixpullrequests.go [hunk 9]
```diff
@@ -307,42 +316,51 @@ func (cfp *CreateFixPullRequestsCmd) updatePackageToFixedVersion(packageType cor
...
 	commandArgs = append(commandArgs, fixedPackage)
-	return runPackageMangerCommand(commandName, commandArgs)
+	return runPackageMangerCommand(technology.GetExecCommandName(), commandArgs)
+}
+
+// Module paths in GO must have a major version suffix like /v2 that matches the major version.
+// For example, if a module has the path example.com/mod at v1.0.0, it must have the path example.com/mod/v2 at version v2.0.0.
+// Also bear in mind, that GitHub uses "/" while gopkg use "." to indicate major version change
+// Further reading https://github.com/golang/go/wiki/Modules#semantic-import-versioning
+func formatGoVersion(impactedPackage, fixVersion string) (string, error) {
```
Comment (sverdlov93) [inline commands/createfixpullrequests.go:351]:
```suggestion
func handleGoPackageSemanticVersionSuffix(impactedPackage, fixVersion string) (string, error) {
```

File: commands/createfixpullrequests.go [hunk 10]
```diff
@@ -307,42 +316,55 @@ func (cfp *CreateFixPullRequestsCmd) updatePackageToFixedVersion(packageType cor
...
+// Module paths in GO must have a major version suffix like /v2 that matches the major version.
+// For example, if a module has the path example.com/mod at v1.0.0, it must have the path example.com/mod/v2 at version v2.0.0.
+// Also bear in mind, that GitHub uses "/" while gopkg use "." to indicate major version change
+// Further reading https://github.com/golang/go/wiki/Modules#semantic-import-versioning
+func handleGoPackageSemanticVersionSuffix(impactedPackage, fixVersion string) (string, error) {
+	pathSeparator := "/"
+	shouldMentionFirstVersion := false // Weather or not to mention /v1 packages
+	majorVersion, err := strconv.Atoi(strings.Split(fixVersion, semanticVersioningSeparator)[0])
+	importPathPrefix := strings.Split(impactedPackage, ".")[0]
+	if err != nil {
```
Comment (sverdlov93) [inline commands/createfixpullrequests.go:356]: line in wrong place

File: commands/createfixpullrequests.go [hunk 11]
```diff
@@ -307,42 +316,55 @@ func (cfp *CreateFixPullRequestsCmd) updatePackageToFixedVersion(packageType cor
...
+		return "", err
+	}
+	if importPathPrefix == gopkg {
+		pathSeparator = "."
+		shouldMentionFirstVersion = true
+	}
+	if majorVersion > 1 || shouldMentionFirstVersion {
+		impactedPackageWithoutVersion := strings.Split(impactedPackage, pathSeparator+"v")[0]
```
Comment (sverdlov93) [inline commands/createfixpullrequests.go:364]: `http://github.com/vbauerster/mpb/v7` will not work in that case
It will split 'http://github.com', 'bauerster/mpb' , 'v7'

Comment (EyalDelarea) [reply commands/createfixpullrequests.go]: Great edge case ! i've fixed it and added a test to verify this cases.
